### PR TITLE
PIM-7263: Purging command for boolean values on variant products

### PR DESCRIPTION
**Pull request by grena**
**Description (for Contributor and Core Developer)**

If variant product has values on boolean attributes that should belong to the parents, this parent will remove those values.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added legacy Behats               | N
| Added acceptance tests            | N
| Added integration tests           | N
| Changelog updated                 | Y
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+ * If some variant products have a some boolean values at their variation level that should belong to their
+ * parents instead, this command will remove these values.
+ *
+ * @author    Adrien Pétremann <adrien.petremann@akeneo.com>
+ * @copyright 2018 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class RemoveWrongBooleanValuesOnVariantProductsCommand extends ContainerAwareCommand
+{
+    const PRODUCT_BULK_SIZE = 100;
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: private const

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: :void

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            ->setHidden(true)
+            ->setName('pim:catalog:remove-wrong-values-on-variant-products')
+            ->setDescription('Remove boolean values on variant products that should belong to their parents')
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: :void

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        foreach ($variantProducts as $variantProduct) {
+            $totalProducts++;
+            $isModified = false;
+
+            foreach ($variantProduct->getFamily()->getAttributes() as $attribute) {
+                if ($this->isProductImpacted($variantProduct, $attribute)) {
+                    $this->cleanProductForAttribute($variantProduct, $attribute);
+                    $isModified = true;
+                }
+
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: extra space

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        return $pqb->execute();
+    }
+
+    /**
+     * @param mixed              $variantProduct
+     * @param AttributeInterface $attribute
+     *
+     * @return bool
+     */
+    private function isProductImpacted($variantProduct, AttributeInterface $attribute): bool
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: mixed? `EntityWithFamilyVariantInterface` ?

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        }
+
+        return null !== $variantProduct->getValuesForVariation()->getByCodes($attribute->getCode());
+    }
+
+    /**
+     * @param mixed              $variantProduct
+     * @param AttributeInterface $attribute
+     */
+    private function cleanProductForAttribute($variantProduct, AttributeInterface $attribute): void
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: idem, mixed ?

**Comment by grena** [comment]
@LaurentPetard We can't wait for monday pull up to bring it to 2.2 (as the original command is asked for 2.2). Therefore, once you'll merge the fix on 2.0, we'll merge the command on 2.0 (the code is ready, we need to create the PR)

File: src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php (hunk 5 above)
Comment (grena) [reply src/Pim/Bundle/CatalogBundle/Command/RemoveWrongBooleanValuesOnVariantProductsCommand.php]: Yes, our `Akeneo\Component\StorageUtils\Cursor\CursorInterface` returns `mixed` entities when we parse it

**Pull request merged**